### PR TITLE
Tests for TPCDS queries join ordering

### PR DESCRIPTION
**Pull request by kokosing**
Tests for TPCDS queries join ordering

### CODE REVIEW

File: presto-tpcds/src/main/resources/tpcds/queries/TODO [hunk 1]
```diff
@@ -0,0 +1 @@
+Move queries to TPCDS generator library.
```
Comment (findepi) [inline presto-tpcds/src/main/resources/tpcds/queries/TODO:1]: We won't find this file. Better to just reg an issue somewhere.

File: presto-tpcds/src/main/resources/tpcds/queries/q01.sql [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+   WHERE ("sr_returned_date_sk" = "d_date_sk")
+      AND ("d_year" = 2000)
+   GROUP BY "sr_customer_sk", "sr_store_sk"
+) 
+SELECT "c_customer_id"
+FROM
+  customer_total_return ctr1
+, store
+, customer
+WHERE ("ctr1"."ctr_total_return" > (
```
Comment (findepi) [inline presto-tpcds/src/main/resources/tpcds/queries/q01.sql:20]: I'm not convinced we need yet another copy of TPCDS queries.
```
$ find -regex '.*ds.*q01.sql' -exec sha1sum {} +
550d3e6497028853792bc46c8500776e92d9971b  ./presto-tpcds/src/main/resources/tpcds/queries/q01.sql
550d3e6497028853792bc46c8500776e92d9971b  ./presto-product-tests/src/main/resources/sql-tests/testcases/tpcds/q01.sql
0169942554f00ce16abf51e98b1730fa8be8e005  ./presto-benchto-benchmarks/src/main/resources/sql/presto/tpcds/q01.sql
```

I would rather see some reuse (even hacky) then a copy.

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseJoinReorderingTest.java [hunk 3]
```diff
@@ -0,0 +1,251 @@
...
+
+import static com.facebook.presto.sql.planner.plan.JoinNode.DistributionType.REPLICATED;
+import static com.facebook.presto.sql.planner.plan.JoinNode.Type.INNER;
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkState;
+import static java.lang.String.format;
+import static java.util.Objects.requireNonNull;
+import static org.testng.Assert.assertEquals;
+
+public class BaseJoinReorderingTest
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseJoinReorderingTest.java:36]: Just to make sure. Is
> Extract BaseJoinReorderingTest

an extract only?

File: presto-tpcds/src/test/resources/join_ordering/09.join_ordering [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+                    cross join:
+                        cross join:
+                            cross join:
+                                cross join:
+                                    cross join:
+                                        cross join:
+                                            cross join:
+                                                cross join:
+                                                    cross join:
+                                                        cross join:
```
Comment (findepi) [inline presto-tpcds/src/test/resources/join_ordering/09.join_ordering:15]: This one looks weird.

File: presto-tpcds/src/test/resources/join_ordering/01.join_ordering [hunk 5]
```diff
@@ -0,0 +1,13 @@
+cross join:
+    join (LEFT, REPLICATED):
+        join (INNER, REPLICATED):
+            tpcds:customer:sf3000.0
+            join (INNER, REPLICATED):
```
Comment (findepi) [inline presto-tpcds/src/test/resources/join_ordering/01.join_ordering:5]: Why are these files needed? I thought the idea was that there is no persistent 'expected' value?

Also, include in the cmt msg how the files were generated (technique, what session settings).

Also, name the files `.txt` to avoid stupid problems like "`.join_ordering` is an unknown extension" or whatever.

File: presto-tpcds/src/main/resources/tpcds/statistics/sf3000/README.txt [hunk 6]
```diff
@@ -0,0 +1,2 @@
+This statistics are copied from from hive connector from TPCDS schema with sf3000.
+So distinct values are not exact and also there are no min nor max values for char-based columns.
```
Comment (findepi) [inline presto-tpcds/src/main/resources/tpcds/statistics/sf3000/README.txt:2]: Include this in the cmt msg too.

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 7]
```diff
@@ -0,0 +1,135 @@
...
+import static com.facebook.presto.sql.analyzer.FeaturesConfig.JoinDistributionType.AUTOMATIC;
+import static com.facebook.presto.sql.analyzer.FeaturesConfig.JoinReorderingStrategy.COST_BASED;
+import static com.google.common.base.Charsets.UTF_8;
+import static java.lang.String.format;
+import static java.util.stream.Collectors.toList;
+import static org.testng.Assert.assertEquals;
+
+/**
+ * This class tests cost-based optimization rules related to joins. It contains unmodified TPCH queries.
+ * This class is using TPCH connector configured in way to mock Hive connector with unpartitioned TPCH tables.
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:45]: Update

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 8]
```diff
@@ -0,0 +1,135 @@
...
+ */
+public class TestTpcdsJoinReordering
+        extends BaseJoinReorderingTest
+{
+    /*
+     * CAUTION: The expected plans here are not necessarily optimal yet. Their role is to prevent
+     * inadvertent regressions. A conscious improvement to the planner may require changing some
+     * of the expected plans, but any such change should be verified on an actual cluster with
+     * large amount of data.
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:54]: Update (assuming the purpose/usage of this test class is different than the purpose of the class dedicated to tpch).

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    protected LocalQueryRunner createQueryRunner(Session session)
+    {
+        return LocalQueryRunner.queryRunnerWithFakeNodeCountForStats(session, 8);
+    }
+
+    @Override
+    protected void createTpchCatalog(LocalQueryRunner queryRunner)
+    {
+        queryRunner.createCatalog(queryRunner.getDefaultSession().getCatalog().get(),
+                new TpcdsConnectorFactory(1),
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:76]: The method is called `createTpchCatalog` ...

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 10]
```diff
@@ -0,0 +1,135 @@
...
+    @Override
+    protected LocalQueryRunner createQueryRunner(Session session)
+    {
+        return LocalQueryRunner.queryRunnerWithFakeNodeCountForStats(session, 8);
+    }
+
+    @Override
+    protected void createTpchCatalog(LocalQueryRunner queryRunner)
+    {
+        queryRunner.createCatalog(queryRunner.getDefaultSession().getCatalog().get(),
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:75]: fmt

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 11]
```diff
@@ -0,0 +1,135 @@
...
+    }
+
+    @DataProvider
+    public Object[][] getTpcdsQueryIds()
+    {
+        return IntStream.range(1, 100)
+                .boxed()
+                .flatMap(i -> {
+                    if (i < 10) {
+                        return Stream.of("0" + i);
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:87]: You can use `format("%02d", i)` below and remove this `if`.

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 12]
```diff
@@ -0,0 +1,135 @@
...
+                    }
+                    if (i == 14 || i == 23 || i == 24 || i == 39) {
+                        return Stream.of(i + "_1", i + "_2");
+                    }
+                    return Stream.of(i.toString());
+                })
+                .map(i -> new Object[]{i})
+                .collect(toList())
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:95]: Directly `.toArray(Object[][]::new);`

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+    public void test(String queryId)
+            throws IOException
+    {
+        assertEquals(joinOrderString(tpcdsQuery(queryId)), read(getExpectedJoinOrderingFile(queryId)));
+    }
+
+    private String read(File file)
+            throws IOException
+    {
+        return Joiner.on("\n").join(Files.readLines(file, UTF_8)) + "\n";
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:109]: Use `Path` instead of `File` and JDK instead of Guava:
```
return java.nio.file.Files.lines(file).collect(joining("\n"));
```

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+    @Test(dataProvider = "getTpcdsQueryIds", enabled = false)
+    public void update(String queryId)
+            throws IOException
+    {
+        Files.write(joinOrderString(tpcdsQuery(queryId)), getExpectedJoinOrderingFile(queryId), UTF_8);
+    }
+
+    private static File getExpectedJoinOrderingFile(String queryId)
+    {
+        return new File(String.format("src/test/resources/join_ordering/%s.join_ordering", queryId));
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:133]: rnn `queryId`

static import `format`

use `Path`

end file name with e.g. `.txt`

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 15]
```diff
@@ -0,0 +1,135 @@
...
+            String queryPath = format("/tpcds/queries/q%s.sql", queryId);
+            URL resourceUrl = Resources.getResource(TpcdsConnectorFactory.class, queryPath);
+            return Resources.toString(resourceUrl, StandardCharsets.UTF_8);
+        }
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test(dataProvider = "getTpcdsQueryIds", enabled = false)
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:124]: Now, calling this requires temporarily removing ` enabled = false`. What about making this a `main` method?

File: presto-tpcds/pom.xml [hunk 16]
```diff
@@ -96,5 +96,17 @@
...
+        <dependency>
+            <groupId>io.airlift</groupId>
+            <artifactId>joni</artifactId>
+            <scope>test</scope>
+        </dependency>
+
+        <dependency>
+            <groupId>io.airlift</groupId>
+            <artifactId>bytecode</artifactId>
+            <scope>test</scope>
```
Comment (findepi) [inline presto-tpcds/pom.xml:109]: interesting, what are these for?

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 17]
```diff
@@ -0,0 +1,135 @@
...
+                .map(i -> new Object[]{i})
+                .collect(toList())
+                .toArray(new Object[][]{});
+    }
+
+    @Test(dataProvider = "getTpcdsQueryIds")
+    public void test(String queryId)
+            throws IOException
+    {
+        assertEquals(joinOrderString(tpcdsQuery(queryId)), read(getExpectedJoinOrderingFile(queryId)));
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:103]: Very nice and functional.

However, if something goes wrong (or one wants to run this with a debugger), this is no longer so nice.

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java [hunk 18]
```diff
@@ -0,0 +1,135 @@
...
+        catch (IOException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    @Test(dataProvider = "getTpcdsQueryIds", enabled = false)
+    public void update(String queryId)
+            throws IOException
+    {
+        Files.write(joinOrderString(tpcdsQuery(queryId)), getExpectedJoinOrderingFile(queryId), UTF_8);
```
Comment (findepi) [inline presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:128]: Very nice and functional.

However, if something goes wrong (or one wants to run this with a debugger), this is no longer so nice.

File: presto-tpcds/src/test/resources/join_ordering/09.join_ordering (hunk 4 above)
Comment (kokosing) [reply presto-tpcds/src/test/resources/join_ordering/09.join_ordering:15]: Yep... this is how we handle this query today, see: https://github.com/prestodb/presto/issues/6944

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestJoinReordering.java [hunk 19]
```diff
@@ -46,18 +37,14 @@
...
-import static com.google.common.base.Preconditions.checkState;
 import static java.lang.String.format;
-import static java.util.Objects.requireNonNull;
-import static org.testng.Assert.assertEquals;
 
 /**
  * This class tests cost-based optimization rules related to joins. It contains unmodified TPCH queries.
  * This class is using TPCH connector configured in way to mock Hive connector with unpartitioned TPCH tables.
  */
 public class TestJoinReordering
```
Comment (findepi) [inline presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/TestJoinReordering.java:46]: `TestTpchJoinReordering`?

File: presto-tpcds/src/main/resources/tpcds/queries/TODO (hunk 1 above)
Comment (kokosing) [reply presto-tpcds/src/main/resources/tpcds/queries/TODO:1]: https://github.com/Teradata/tpcds/issues/28

Also I asked on presto slack to copy that project under `prestodb` community group, see https://prestodb.slack.com/archives/C07JH9WMQ/p1519824551000061

File: presto-tpcds/src/main/resources/tpcds/queries/q01.sql (hunk 2 above)
Comment (kokosing) [reply presto-tpcds/src/main/resources/tpcds/queries/q01.sql:20]: queries in benchmarks are modified (they have variables like `${catalog}.${schema}`).
Product tests have their headers.  Here we need vanilla queries.

I prefer to copy files which no-one ever should modify, instead of dragging and/or playing with all the product-tests deps (like cassandra or kafka).

File: presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseJoinReorderingTest.java (hunk 3 above)
Comment (kokosing) [reply presto-main/src/test/java/com/facebook/presto/sql/planner/optimizations/BaseJoinReorderingTest.java:36]: Sure it is. :)

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java (hunk 8 above)
Comment (kokosing) [reply presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:54]: I am not sure if the purpose is different. Practically it works in the same way.

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java (hunk 12 above)
Comment (kokosing) [reply presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:95]: refactored

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java (hunk 15 above)
Comment (kokosing) [reply presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:124]: Then it has to be copied to both classes ; /

File: presto-tpcds/src/main/resources/tpcds/queries/q01.sql (hunk 2 above)
Comment (findepi) [reply presto-tpcds/src/main/resources/tpcds/queries/q01.sql:20]: > queries in benchmarks are modified (they have variables like ${catalog}.${schema}).

you could simply find-replace `${catalog}` with `tpcds` ...

> I prefer to copy files which no-one ever should modify,

This is wishful thinking. The queries _are_ being modified from time to time. Casts, grouping, ..  And they will be modified in the future (casts and chars).

File: presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java (hunk 15 above)
Comment (findepi) [reply presto-tpcds/src/test/java/com/facebook/presto/sql/planner/optimizations/TestTpcdsJoinReordering.java:124]: > Then it has to be copied to both classes ; /

this i don't understand

**Comment by kokosing** [comment]
Superseded by https://github.com/starburstdata/presto/pull/29
